src/main/java/org/sonarlint/intellij/issue/persistence/IndexedObjectStore.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * SonarLint for IntelliJ IDEA
+ * Copyright (C) 2015 SonarSource
+ * sonarlint@sonarsource.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonarlint.intellij.issue.persistence;
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.vfs.VirtualFile;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Optional;
+import org.sonarsource.sonarlint.core.client.api.connected.objectstore.ObjectStore;
+import org.sonarsource.sonarlint.core.client.api.connected.objectstore.PathMapper;
+import org.sonarsource.sonarlint.core.client.api.connected.objectstore.Reader;
+import org.sonarsource.sonarlint.core.client.api.connected.objectstore.Writer;
+
+/**
+ * An ObjectStore without internal cache that derives the filesystem path to storage using a provided PathMapper.
+ *
+ * @param <K> type of the key to store by and used when reading back; must be hashable
+ * @param <V> type of the value to store
+ */
+public class IndexedObjectStore<K, V> implements ObjectStore<K, V> {
+  private static final Logger LOGGER = Logger.getInstance(IndexedObjectStore.class);
+  private final StoreIndex<K> index;
+  private final PathMapper<K> pathMapper;
+  private final Reader<V> reader;
+  private final Writer<V> writer;
+  private final StoreKeyValidator<K> validator;
+
+  public IndexedObjectStore(StoreIndex<K> index, PathMapper<K> pathMapper, Reader<V> reader, Writer<V> writer, StoreKeyValidator<K> validator) {
+    this.index = index;
+    this.pathMapper = pathMapper;
+    this.reader = reader;
+    this.writer = writer;
+    this.validator = validator;
+  }
+
+  @Override
+  public Optional<V> read(K key) throws IOException {
+    Path path = pathMapper.apply(key);
+    if (!path.toFile().exists()) {
+      return Optional.empty();
+    }
+    return Optional.of(reader.apply(Files.newInputStream(path)));
+  }
+
+  /**
+   * Deletes all entries in the index are no longer valid.
+   */
+  public void clean() {
```

Comment:
The meaning of `clean` was not obvious to me when reading a usage, for example in a test case. How about `deleteInvalid` ? It would nicely make the logical connection with the validator and deletion.